Add utility class SortBuffer

kewang1024 opened the pull request:
SortBuffer is a utility class to accumulate data inside and output the sorted result.
Spilling would be triggered if spilling is enabled and memory usage exceeds limit.

kewang1024 commented:
Reference PR: https://github.com/facebookincubator/velox/pull/2259

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,131 @@
+  std::unique_ptr<Spiller> sortedSpiller_;
+  // Read back the spilled data.
+  std::unique_ptr<TreeOfLosers<SpillMergeStream>> spillMerger_;
+  // Record the source rows to copy to 'output_' in order.
```
```
// Records the source ...
```

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,131 @@
+      std::optional<Spiller::Config> spillConfig = std::nullopt,
+      uint64_t spillMemoryThreshold = 0);
+
+  void add(const RowVectorPtr& input);
```
```
s/add/addInput/
```

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,131 @@
+  /// We break the result into batches, each call of output() return one batch
+  /// of result, output() returning nullptr indicates all results have been
+  /// returned.
+  RowVectorPtr output();
```
```
s/output/getOutput/
```

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,131 @@
+
+  void add(const RowVectorPtr& input);
+
+  /// Indicate no more input. After it's been called, add() can't be called on
```
```
s/Indicate/Indicates/
```

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,131 @@
+
+  /// API for Abitrator to recycle memory from this object.
+  /// Will trigger spilling and free targetBytes of memory.
+  /// @param targetBytes target memory bytes to reclaim.
```
Let's not use @param this is usually used for public API but not for internal function annotations.

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,131 @@
+  void reclaim(uint64_t targetBytes);
+
+ private:
+  void triggerSpillIfNeeded(const RowVectorPtr& input);
```
Comment these methods and leave an empty line in between?

xiaoxmeng commented on velox/exec/SortBuffer.cpp:
```diff
@@ -0,0 +1,348 @@
+
+void SortBuffer::add(const RowVectorPtr& input) {
+  VELOX_CHECK(!noMoreInput_);
+  triggerSpillIfNeeded(input);
```
Let's just call it ensureInput

xiaoxmeng commented on velox/exec/SortBuffer.cpp:
```diff
@@ -0,0 +1,348 @@
+  }
+
+  // For variable length data, we take the flat size of the input as the cap.
+  const int64_t estimatedIncrementalBytes =
```
Please use the original variable names in order by to keep the code the same. You can do cleanup and we can review that part separately

xiaoxmeng commented on velox/exec/SortBuffer.cpp:
```diff
@@ -0,0 +1,348 @@
+    return;
+  }
+
+  // 1. If current memory usage exceeds spilling threshold, trigger spilling.
```
Do not number the steps inside function.

xiaoxmeng commented on velox/exec/SortBuffer.cpp:
```diff
@@ -0,0 +1,348 @@
+
+  // Return data in batches
+  initOrResetOutputBuffer();
+  if (sortedSpiller_ == nullptr) {
```
I think the original code is more clear to me
```
  prepareOutput();

  if (spiller_ != nullptr) {
    getOutputWithSpill();
  } else {
    getOutputWithoutSpill();
  }
```

xiaoxmeng reviewed (commented):
@kewang1024 thanks for the change. I saw many rename on variable and function names in this change. Please keep the original code in order by. We can do a code cleanup if necessary.

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,159 @@
+
+namespace facebook::velox::functions::test {
+
+class SortBufferTest : public testing::Test,
```
For tests, let's cover all the existing tests in OrderByTest. Thanks!

kewang1024 commented on velox/exec/SortBuffer.cpp:
```diff
@@ -0,0 +1,348 @@
+    velox::memory::MemoryPool* memoryPool,
+    tsan_atomic<bool>* nonReclaimableSection,
+    uint32_t outputBatchSizeLimit,
+    std::optional<Spiller::Config> spillConfig,
```
In OrderBy, it's `std::optional<Spiller::Config>` , so when passing it down, it's better to keep it consistent

xiaoxmeng replied:
Let's pass it as nullptr to ease the handling here?

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,142 @@
+ public:
+  SortBuffer(
+      const RowTypePtr& input,
+      const std::vector<column_index_t>& sortedIndex,
```
```
s/sortedIndex/sortColumnIndices/
```

xiaoxmeng replied:
@kewang1024 could you fix? thanks!

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,142 @@
+  SortBuffer(
+      const RowTypePtr& input,
+      const std::vector<column_index_t>& sortedIndex,
+      const std::vector<CompareFlags>& sortedProperties,
```
```
s/sortedProperties/sortCompareFlags/
```

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,142 @@
+      tsan_atomic<bool>* nonReclaimableSection,
+      uint32_t outputBatchSizeLimit_ = 10000,
+      std::optional<Spiller::Config> spillConfig = std::nullopt,
+      uint64_t spillMemoryThreshold = 0);
```
Can you move spillMemoryThreshold into Spiller::Config in followup? Thanks!

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,142 @@
+
+  void addInput(const RowVectorPtr& input);
+
+  /// Indicates no more input. After it's been called, addInput() can't be
```
Consider
```
  /// Indicates no more input which either triggers in-memory sorting on rows stored 'data_' if spilling is not triggered, or finish the spilling and setup the sort merge reader to prepare for the un-spilling processing for output.
```

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,142 @@
+  ///  - finish the spilling and setup merge reader to prepare un-spilling
+  std::optional<SpillStats> noMoreInput();
+
+  /// We break the result into batches, each call of getOutput() return one
```
Let's make the comment concise.
```
/// Returns the sorted output rows in batch.
```

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,142 @@
+
+  /// API for Abitrator to recycle memory from this object.
+  /// Will trigger spilling and free targetBytes of memory.
+  void reclaim(uint64_t targetBytes);
```
Just expose and let operator handle reclaim API
```
void spill(int64_t targetRows, int64_t targetBytes);
```
We can check spill config has been setup in spill

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,142 @@
+  ///  - current memory usage exceeds spilling threshold
+  ///  - current available reservation in memory pool is not enough and
+  ///  furthermore we fail to reserve more in memory pool
+  void ensureInputFits(const RowVectorPtr& input);
```
Let's keep the comment concise
```
// Ensures there is sufficient memory reserved to process 'input'.
void ensureInputFits(const RowVectorPtr& input);
```
We might change the implementation later.

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,142 @@
+  ///  - current available reservation in memory pool is not enough and
+  ///  furthermore we fail to reserve more in memory pool
+  void ensureInputFits(const RowVectorPtr& input);
+  /// Spill targetRows/targetBytes from 'data_' to disk to free some memory.
```
Private method use //
```
/// Invoked to spill targetRows/targetBytes from 'data_' to disk to reclaim memory for the threshold based disk spilling trigger or memory arbitration.
```

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,142 @@
+  void spill(int64_t targetRows, int64_t targetBytes);
+  /// Initialize the reusable output buffer, if it's initialized already, reset
+  /// it for next iteration.
+  void initOrResetOutputBuffer();
```
Please keeps the original method name in order by. We just need to highlight what a function does.

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,142 @@
+  // To use the sorting facility provided by the row container, store
+  // the data in example structures above: (c2, c1, c0, c3)
+  std::shared_ptr<RowContainer> data_;
+  // The flag is passed from the associated operator such as order by or table
```
The comment is irrelevant

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,142 @@
+  void prepareOutput();
+  void prepareOutputWithSpill();
+
+  // Input/output data layout properties
```
Let's make the comment concise.
```
  const RowTypePtr input_;
  const std::vector<CompareFlags> sortedProperties_;
  memory::MemoryPool* const pool_;
  // The maximum output batch size in rows.
  const uint32_t outputBatchSize_;

  // The column projection map for sort columns between 'input_' and 'spillerStoreType_' as sort buffer stores the sort columns first in 'data_'.
  std::vector<IdentityProjection> columnMap_;

  // Indicates no more input. Once it is set, addInput() can't be called on this sort buffer object.
  bool noMoreInput_;
  // The number of received input rows.
  size_t numInputRows_{0};
  // Used to store the input data in row format.
  std::shared_ptr<RowContainer> data_;

  std::vector<char*> sortedRows_;

  // The flag is passed from the associated operator such as order by or table writer to indicate if this sort buffer object is under non-reclaimable execution section or not.
  tsan_atomic<bool>* const nonReclaimableSection_;

  // Spilling related state:
  // If not null, indicates spilling is enabled.
  const std::optional<Spiller::Config> spillConfig_;
  // The maximum size of data that the sort buffer can hold in memory before spilling. If it zero, then there is no limit.
  //
  // NOTE: this only applies if disk spilling is enabled.
  const uint64_t spillMemoryThreshold_;
  // The data type of the rows stored in 'data_' and spilled on disk. The
  // sort key columns are stored first then the non-sorted data columns.
  std::unique_ptr<Spiller> spiller_;
  // Used to merge the sorted runs from in-memory rows and spilled rows on disk.
  std::unique_ptr<TreeOfLosers<SpillMergeStream>> spillMerger_;
  // Records the source rows to copy to 'output_' in order.
  std::vector<const RowVector*> spillSources_;
  std::vector<vector_size_t> spillSourceRows_;

  // Counts input batches to trigger spilling for test.
  uint64_t testingSpillCounter_{0};

  // Reusable output vector.
  RowVectorPtr output_;
  // The number of rows that has been returned.
  size_t numOutputRows_{0};
```

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,160 @@
+  ASSERT_EQ(sortBuffer->getOutput(), nullptr);
+}
+
+TEST_F(SortBufferTest, withSpilling) {
```
```
DISABLED_withSpilling
```

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,160 @@
+      spillMemoryThreshold);
+
+  vector_size_t numRows = 35000; // batchSizeLimit_ is 10000, thus 4 batches.
+  const std::shared_ptr<memory::MemoryPool> fuzzerPool_ =
```
```
s/fuzzerPool_/fuzzerPool/
```

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,160 @@
+      {true, true, false, false},
+      {true, true, false, false}};
+
+  const std::shared_ptr<memory::MemoryPool> pool_ =
```
The default memory pool doesn't have memory usage tracking so you first need to create a root memory pool and then create a leaf pool from it. See the example in HashTableTest. Thanks!

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,115 @@
+  /// Returns the sorted output rows in batch.
+  RowVectorPtr getOutput();
+
+  /// Invoked to spill targetRows/targetBytes from 'data_' to disk and free
```
```
/// Invoked to spill with specified targets from 'data_' ...
///
/// NOTE: if either 'targetRows' or 'targetBytes' is zero, then we spill all the rows from 'data_'.
```

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,115 @@
+
+  void addInput(const RowVectorPtr& input);
+
+  /// Indicates no more input and triggers either of:
```
```
s/rows stored/rows in stored/
```

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,115 @@
+  // Ensures there is sufficient memory reserved to process 'input' by spilling.
+  // If spilling is not enabled, no action.
+  void ensureInputFits(const RowVectorPtr& input);
+  // Initialize the reusable output buffer, if it's initialized already, reset
```
```
// Invoked to initialize or reset the reusable output buffer to get output.
```

xiaoxmeng commented on velox/exec/SortBuffer.cpp:
```diff
@@ -0,0 +1,360 @@
+
+  std::vector<TypePtr> sortedColumnTypes;
+  std::vector<TypePtr> nonSortedColumnTypes;
+  std::vector<std::string> sortedSpillNames;
```
```
s/sortedSpillNames/sortedSpillColumnNames/
s/sortedSpillTypes/sortedSpillColumnTypes/
```

xiaoxmeng commented on velox/exec/SortBuffer.cpp:
```diff
@@ -0,0 +1,360 @@
+    velox::memory::MemoryPool* pool,
+    tsan_atomic<bool>* nonReclaimableSection,
+    uint32_t outputBatchSize,
+    std::unique_ptr<Spiller::Config> spillConfig,
```
Let's use
```
Spiller::Config* spillConfig
```

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,114 @@
+  RowVectorPtr getOutput();
+
+  /// Invoked to spill from 'data_' to disk with specified targets.
+  /// NOTE: if either 'targetRows' or 'targetBytes' is zero, then we spill all
```
Leave an empty line between comment and note. Thanks!

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,114 @@
+
+  const RowTypePtr input_;
+  const std::vector<CompareFlags> sortCompareFlags_;
+  const Spiller::Config* spillConfig_;
```
```
const Spiller::Config* const spillConfig_;
```

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,114 @@
+  const Spiller::Config* spillConfig_;
+  // The maximum size that an SortBuffer can hold in memory before spilling.
+  // Zero indicates no limit.
+  // NOTE: 'spillMemoryThreshold_' only applies if disk spilling is enabled.
```
ditto

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,114 @@
+  // Zero indicates no limit.
+  // NOTE: 'spillMemoryThreshold_' only applies if disk spilling is enabled.
+  const uint64_t spillMemoryThreshold_;
+  // Maximum number of rows in the output batch.
```
```
// Maximum number of rows to return in one output batch.
```

xiaoxmeng commented on velox/exec/SortBuffer.h:
```diff
@@ -0,0 +1,114 @@
+  // The column projection map between 'input_' and 'spillerStoreType_' as sort
+  // buffer stores the sort columns first in 'data_'.
+  std::vector<IdentityProjection> columnMap_;
+  velox::memory::MemoryPool* pool_;
```
```
velox::memory::MemoryPool* const pool_;
```
move the const (const pointer) to const members.

xiaoxmeng commented on velox/exec/SortBuffer.cpp:
```diff
@@ -0,0 +1,360 @@
+}
+
+std::optional<SpillStats> SortBuffer::noMoreInput() {
+  VELOX_CHECK(!noMoreInput_);
```
I assume there is no logic change in these function? thanks!

kewang1024 replied:
Correct

xiaoxmeng commented on velox/exec/SortBuffer.cpp:
```diff
@@ -0,0 +1,360 @@
+    return;
+  }
+
+  const int64_t currentDataRows = data_->numRows();
```
```
s/currentDataRows/numRows/
```

xiaoxmeng commented on velox/exec/SortBuffer.cpp:
```diff
@@ -0,0 +1,360 @@
+}
+
+void SortBuffer::spill(int64_t targetRows, int64_t targetBytes) {
+  VELOX_CHECK_GE(targetRows, 0);
```
As you mentioned offline, we could add spillConfig_ != nullptr check here:
```
VELOX_CHECK_NOTNULL(spillConfig_);
if (spiller_ == nullptr) {
  ...
}
...
```

kewang1024 replied:
I gave it a second thought and think I need to confirm with you, what is the expected behavior for reclaim function if spilling is not enabled
- throw exception? or
- do nothing?

xiaoxmeng commented on velox/exec/SortBuffer.cpp:
```diff
@@ -0,0 +1,360 @@
+    // TODO: support fine-grain memory recycling after row container memory
+    // compaction support. As of now, we only support releasing all the memory
+    // when we spill everything.
+    if (targetRows == 0 || targetBytes == 0) {
```
The condition should be
```
if ((data_->numRows() == 0) {
   data_->clear();
}
```
Let's call pool release in reclaim method in operator instead of here.

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,163 @@
+class SortBufferTest : public testing::Test,
+                       public velox::test::VectorTestBase {
+ public:
+  // c0,         c1,         c2,         c3,         c4,         c5
```
Remove the comment here, and comment sortColumnIndices_
```
// Specifies the sort columns ["c4", "c1"].
std::vector<column_index_t> sortColumnIndices_{4, 1};
```

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,163 @@
+  // c0,         c1,         c2,         c3,         c4,         c5
+  // BIGINT()    INTEGER()   SMALLINT()  REAL()      DOUBLE()    VARCHAR()
+  //             sorted-2                            sorted-1
+  RowTypePtr inputType_ = ROW(
```
All these seem to be const members? Thanks!

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,163 @@
+TEST_F(SortBufferTest, batchOutput) {
+  auto sortBuffer = new SortBuffer(
+      inputType_,
+      sortColumnIndices_,
```
Shall we explicit set the output batch size and please remove the default value for output batch size input for sort buffer?

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,163 @@
+      sortColumnIndices_,
+      sortCompareFlags_,
+      pool_.get(),
+      &nonReclaimableSection_);
```
I think it might be better to structure the test as
```
struct {
   bool triggerSpill;
   size_t numInputRows;
   size_t outputBatchSize; -> test 1 such special case. Also please add CHECK in sort buffer ctor, outputBatchSize can't be zero > 0
  std::vector<size_t> expectedOutputBufferSizes;
} testSettings[] = { => list different test scenarios and check existing test cases for example
```

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,163 @@
+      &nonReclaimableSection_);
+
+  vector_size_t numRows = 35000;
+  RowVectorPtr data = makeRowVector(
```
Let's use the fuzzer to generate the vectors as the test below?

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,163 @@
+    ASSERT_NE(output, nullptr);
+  }
+  ASSERT_EQ(sortBuffer->getOutput(), nullptr);
+}
```
I think it is better to have a test utility here
```
RowVectorPtr buildExpectedSortResult(std::vector<> inputVectors, sort flags)
```
Then use this to verify the results that we yield from the sort buffer result?

kewang1024 replied:
but what is the implementation of this test utility? there is no utilities we can use in velox that can do sorting for inputVectors by sort keys?

xiaoxmeng reviewed (commented):
@kewang1024 thanks for the update!

kewang1024 commented:
@xiaoxmeng thanks for the review, addressed some of the comments, haven't finished the test yet, want to confirm, the test regarding reclaim memory would be still in OrderBy test instead of SortBuffer test, is that correct?

xiaoxmeng commented on velox/exec/SortBuffer.cpp:
```diff
@@ -0,0 +1,364 @@
+  VELOX_CHECK(!noMoreInput_);
+  ensureInputFits(input);
+
+  // Prevents the memory arbitrator to reclaim memory from this SortBuffer
```
```
nit: s/SortBuffer/sort buffer/
```

xiaoxmeng commented on velox/exec/SortBuffer.cpp:
```diff
@@ -0,0 +1,364 @@
+    return nullptr;
+  }
+
+  // Return data in batches
```
Remove // Return data in batches? Thanks!

xiaoxmeng commented on velox/exec/SortBuffer.cpp:
```diff
@@ -0,0 +1,364 @@
+    VELOX_CHECK_EQ(spiller_->state().maxPartitions(), 1);
+  }
+
+  // Internally, It would sort and then spill.
```
Remove 
// Internally, It would sort and then spill.

xiaoxmeng reviewed (commented):
@kewang1024 thanks for the update. The production code part looks good % a few nites.

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,356 @@
+using namespace facebook::velox;
+using namespace facebook::velox::memory;
+
+namespace {
```
Move 
```
namespace facebook::velox::functions::test {
namespace {
```

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,356 @@
+    int expectedResultIndex) {
+  ASSERT_GT(output->size(), 0);
+  ASSERT_GE(expectedResult->size(), output->size());
+  for (int i = 0; i < output->size(); ++i) {
```
You can compare one row at a time here? thanks!

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,356 @@
+      {true, true, false, false},
+      {true, true, false, false}};
+
+  const int64_t kMaxBytes = 20LL << 20; // 20 MB
```
Move these fields to protected section
```
s/kMaxBytes/maxBytes_/
```

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,356 @@
+  const std::shared_ptr<memory::MemoryPool> rootPool_{
+      memory::defaultMemoryManager().addRootPool("SortBufferTest", kMaxBytes)};
+  const std::shared_ptr<memory::MemoryPool> pool_{
+      rootPool_->addLeafChild("SortBufferTest", kMaxBytes)};
```
20MB is too small for a memory pool? Set it to 500MB by default and you can create a custom memory pool with small capacity for testing purpose.

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,356 @@
+}
+
+// --run-disabled
+TEST_F(SortBufferTest, DISABLED_dataLayout) {
```
Why you call this data layout?

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,356 @@
+
+TEST_F(SortBufferTest, batchOutput) {
+  struct {
+    std::vector<size_t> numInputRows;
```
Can you add to test if spilling is enabled with test pct injection?

xiaoxmeng reviewed (commented):
@kewang1024 thanks for the update!

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,356 @@
+namespace {
+RowVectorPtr buildExpectedSortResult(std::vector<RowVectorPtr> inputVectors) {}
+
+void verifyData(
```
Leave this function empty with a TODO

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,356 @@
+  }
+}
+
+TEST_F(SortBufferTest, spill) {
```
Can you verify the spilling stats by following similar examples in SpillerTest.cpp? Thanks!
```
  explicit SpillerTest(const TestParam& param)
      : param_(param),
        type_(param.type),
        executorPoolSize_(param.poolSize),
        compressionKind_(param.compressionKind),
        hashBits_(0, type_ == Spiller::Type::kOrderBy ? 0 : 2),
        numPartitions_(hashBits_.numPartitions()),
        statWriter_(std::make_unique<TestRuntimeStatWriter>(stats_)) {
    setThreadLocalRunTimeStatWriter(statWriter_.get());
  }

  ~SpillerTest() {
    setThreadLocalRunTimeStatWriter(nullptr);
  }
```

xiaoxmeng commented on velox/common/base/CompareFlags.h:
```diff
@@ -43,4 +43,16 @@ struct CompareFlags {
+    case CompareFlags::NullHandlingMode::StopAtNull:
+      return "StopAtNull";
+    case CompareFlags::NullHandlingMode::StopAtRhsNull:
+      return "StopAtRhsNull";
```
```
s/NullHandlingModeToStr/nullHandlingModeToStr/
```
Add default? And move implementation to cc?

kewang1024 replied:
There is only a .h file

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,424 @@
+
+namespace facebook::velox::functions::test {
+namespace {
+std::string toString(std::vector<CompareFlags> sortCompareFlags) {
```
How about add toString for CompareFlags? Thanks!

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,424 @@
+
+  folly::Random::DefaultGenerator rng_;
+
+ protected:
```
Remove second protected. And move class members after methods? Thanks!

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,424 @@
+          expectedOutputBufferSizesStr.str());
+    }
+  } testSettings[] = {
+      {false, {2000}, 10000, {2000}},
```
Let's test each case in spilling and non-spilling modes both?

xiaoxmeng reviewed (commented):
@kewang1024 thanks for iterating. Overall looks good % minors. Thanks!

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,424 @@
+  } testSettings[] = {
+      {false, {2000}, 10000, {2000}},
+      {false, {2000}, 2000, {2000}},
+      {true, {1024, 1024, 1024}, 1000, {1000, 1000, 1000, 72}}};
```
Do we want to cover output batch size of 1 such special case? Thanks!

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,424 @@
+      ASSERT_FALSE(spillStats.has_value());
+    } else {
+      ASSERT_TRUE(spillStats.has_value());
+      ASSERT_GT(spillStats->spilledRows, 0);
```
```
ASSERT_EQ(spillStats->spilledRows, numInput) as we discussed offline?
```

xiaoxmeng commented on velox/common/base/CompareFlags.h:
```diff
@@ -41,6 +43,24 @@ struct CompareFlags {
+      case CompareFlags::NullHandlingMode::StopAtNull:
+        return "StopAtNull";
+      case CompareFlags::NullHandlingMode::StopAtRhsNull:
+        return "StopAtRhsNull";
```
```
static std::string nullHandlingModeToStr(NullHandlingMode mode) {
   switch (mode) {
default:
   return fmt::format("Unknown Null Handling mode {}", static_cast<int>(nullHandlingMode))
```

xiaoxmeng commented on velox/common/base/CompareFlags.h:
```diff
@@ -41,6 +43,24 @@ struct CompareFlags {
+
+  std::string toString() {
+    std::stringstream sortCompareFlagsStr;
+    sortCompareFlagsStr << nullsFirst << "," << ascending << "," << equalsOnly
```
```
return fmt::format("[NullFirst[{}] Ascending[{}] EqualsOnly[{}] NullHandleMode[{}]]", ...);
```

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,419 @@
+      memory::defaultMemoryManager().addRootPool("SortBufferTest", maxBytes_)};
+  const std::shared_ptr<memory::MemoryPool> pool_{
+      rootPool_->addLeafChild("SortBufferTest", maxBytes_)};
+  std::shared_ptr<folly::Executor> executor_{
```
```
const std::shared_ptr<folly::Executor> executor_{
```

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,419 @@
+  ASSERT_EQ(output->childAt(1)->asFlatVector<int32_t>()->valueAt(4), 2);
+}
+
+// --run-disabled
```
```
// TODO: enable it later with test utility to compare the sorted result.
```

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,419 @@
+    std::vector<int32_t> expectedResult;
+
+    std::string debugString() const {
+      std::stringstream expectedResultStr;
```
```
const std::string expectedResultStr = folly::join(",", expectedResult);
```

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,419 @@
+          expectedResult.end(),
+          std::ostream_iterator<size_t>(expectedResultStr, " "));
+      std::stringstream sortCompareFlagsStr;
+      for (auto sortCompareFlag : sortCompareFlags) {
```
```
const auto& sortCompareFlag
```

xiaoxmeng commented on velox/exec/tests/SortBufferTest.cpp:
```diff
@@ -0,0 +1,419 @@
+          std::ostream_iterator<size_t>(expectedResultStr, " "));
+      std::stringstream sortCompareFlagsStr;
+      for (auto sortCompareFlag : sortCompareFlags) {
+        sortCompareFlagsStr << sortCompareFlag.toString() << ";";
```
```
s/;/./
```

xiaoxmeng reviewed (approved):
@kewang1024 LGTM. Thanks for iterating!
